### PR TITLE
Enh/split sequences

### PR DESCRIPTION
**Pull request by MichKe95**
Changes proposed in this pull request:
- Segment can now be provided resolve_overlap_flag=True which results in the segment resolving overlapping elements by finding the smallest set of mutually overlapping elements, putting all pulses in a new element and resolve all timings, trigger pulses for this new element
- we achieve this by introducing an optional step (resolve_overlap() ) in the resolve_segment() routine after having added all pulses and elements and before the trigger pulses are generated
- this routine first determines which elements are overlapping before combining them into one element using _combine_elements()
- after this step, the trigger pulses are generated for the resolved set of elements taking care of the different hardware requirements

Conducted Tests on virtual setup:
- randomly over lapping sequences get sorted properly into different elements that all mutually overlap
- Rabi sequence works nominally (to be expected)
- Ramsey two cases: a) with spacing where the two pi pulses overlap b) with spacing where they do not overlap. In both cases the routine did resolve the overlap correctly (including trigger pulses)

Next step:
- test the code on BF1 setup
- run Ramsey style sequences as in the virtual case
- make sure no weird timing effects occur (visible in Ramsey if timing is off) 

Potential improvements:
- change resolve_overlap_flag to be a list of AWGs for which this rule should apply.

Once the hardware tests are done I invite @antsr (and either @chellings or @nathlacroix ) for review.

### CODE REVIEW

File: pycqed/measurement/waveform_control/segment.py [hunk 1]
```diff
@@ -67,12 +73,17 @@ def __init__(self, name, pulse_pars_list=[]):
...
+        self.overlapping_elements = []
         self.timer = Timer(self.name)
         self.pulse_pars = []
 
         for pulse_pars in pulse_pars_list:
             self.add(pulse_pars)
 
+        self.resolve_overlapping_elements = \
+            kw.pop('resolve_overlapping_elements',
+                   self.pulsar.resolve_overlapping_elements())
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py:108]: undocumented kw

File: pycqed/measurement/waveform_control/segment.py [hunk 2]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
                 if (el_prev_end - el_new_start) > tol :
-                    print(el_prev_end, el_new_start)
-                    msg = f'{prev_el} (ends at {el_prev_end*1e6:.4f}us) and ' \
-                    f'{el_list[i + 1][2]} (' \
-                        f'starts at {el_new_start*1e6:.4f}us) overlap ' \
-                          f'on {awg}'
-                    if allow_overlap:
-                        log.warning(msg)
+                    if track_and_ignore:
+                        # add set of two overlapping elements to list
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py:862]: Have you tested whether your code works correctly in case of more than two overlapping elements? Looking at the criterion `(el_prev_end - el_new_start) > tol`, I wonder whether there are cases where an overlapped might stay undetected (e.g., a long element starting first plus two short non-overlapping elements that both overlap with the long element).

File: pycqed/measurement/waveform_control/segment.py [hunk 3]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+                              f'on {awg}'
+                        if allow_overlap:
+                            log.warning(msg)
+                        else:
+                            raise ValueError(msg)
+
+    def resolve_overlap(self):
+        """
+        Routine to resolve overlapping elements. Will be exectued if
+        corresponding pulsar parameter is true. This code
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py]: Will be executed if self.resolve_overlapping_elements is True.

File: pycqed/measurement/waveform_control/segment.py [hunk 4]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+        joint_overlapping_elements = [overlapping_elements[0]]
+
+        new_cluster = True
+        for i in range(len(overlapping_elements) - 1):
+            # check for all sets added to joint_overlapping_elements
+            # whether the next set of elements from
+            # overlapping_elements shares an element name. If so,
+            # then add this element to the entry in
+            # joint_overlapping_elements
+            for j in range(len(joint_overlapping_elements)):
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py]: Unless I am missing something, the check via this loop is more complicated than needed. overlapping_elements should be in correct order due to `el_list.sort()` in `_test_overlap`, so is it really needed to check all entries of `joint_overlapping_elements` here?

File: pycqed/measurement/waveform_control/segment.py [hunk 5]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+        pairwise overlapping and clusters them into lists
+        of overlapping elements, where two different lists of overlapping
+        elements have no overlapping elements with
+        one another. At the end the code combines all elements of each
+        list into a new element.
+        """
+
+        self.gen_elements_on_awg()
+        self._test_overlap(track_and_ignore=True)
+        overlapping_elements = self.overlapping_elements
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py]: Why is `self.overlapping_elements` a property? It could just be return value of `_test_overlap` if that method is called with `track_and_ignore=True`.

If there is a reason to have it as a property: should the property be updated to an empty list after resolving the overlaps?

File: pycqed/measurement/waveform_control/segment.py [hunk 6]
```diff
@@ -934,6 +1034,10 @@ def element_start_length(self, element, awg):
             t_start = min(pulse.algorithm_time(), t_start)
             t_end = max(pulse.algorithm_time() + pulse.length, t_end)
 
+        if t_start == float('inf') or t_end == -float('inf'):
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py]: Is there a reason against using np.inf?

Please add a comment explaining in which instances we expect to end up in this branch of code and why it can be silently ignored.

File: pycqed/measurement/waveform_control/segment.py [hunk 7]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+            #remove it from element_start_end
+            self.element_start_end.pop(el)
+
+        # add new element
+        self.elements[combined_el_name] = new_pulse_list
+        # update new elements_on_awg
+        self.gen_elements_on_awg()
+        # update element_start_end
+
+        #for awg in self.pulsar.awgs:
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py]: remove commented code

File: pycqed/measurement/waveform_control/segment.py [hunk 8]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+            new_pulse_list+= self.elements.pop(el)
+            #remove it from element_start_end
+            self.element_start_end.pop(el)
+
+        # add new element
+        self.elements[combined_el_name] = new_pulse_list
+        # update new elements_on_awg
+        self.gen_elements_on_awg()
+        # update element_start_end
+
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py:958]: please keep explaining comments together with the corresponding code, i.e., the blank line should be above this comment, not below.

File: pycqed/measurement/waveform_control/segment.py [hunk 9]
```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+        Routine to properly combine elements in the segment.
+        :param elements: list or set of elements in the segment to be combined
+        :param combined_el_name: name of the combined element
+        :return:
+        """
+
+        new_pulse_list = []
+
+        for el in elements:
+            new_pulse_list+= self.elements.pop(el)
```
Comment (chellings) [inline pycqed/measurement/waveform_control/segment.py]: Shouldn't you update `p.pulse_obj.element_name` to the new name for p in new_pulse_list?

**Review by chellings — CHANGES_REQUESTED** [review]
Thanks for this useful feature. Please see my comments.

Please make sure to test the recent and final changes on a virtual setup. Or do you think that another test on a real setup will be needed?

File: pycqed/measurement/waveform_control/segment.py [hunk 10]
```diff
@@ -810,15 +836,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
-                          f'on {awg}'
-                    if allow_overlap:
-                        log.warning(msg)
+                    if track_and_ignore:
+                        # add set of two overlapping elements to list
+                        self.overlapping_elements.append({prev_el,
+                                                          el_list[i + 1][2]})
                     else:
-                        raise ValueError(msg)
+                        print(el_prev_end, el_new_start)
```
Comment (antsr) [inline pycqed/measurement/waveform_control/segment.py]: remove print statement

File: pycqed/measurement/waveform_control/segment.py [hunk 11]
```diff
@@ -810,15 +836,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+        elements have no overlapping elements with
+        one another. At the end the code combines all elements of each
+        list into a new element.
+        """
+
+        self.gen_elements_on_awg()
+        self._test_overlap(track_and_ignore=True)
+        overlapping_elements = self.overlapping_elements
+
+        if len(overlapping_elements)==0:
```
Comment (antsr) [inline pycqed/measurement/waveform_control/segment.py]: spaces around `==`

File: pycqed/measurement/waveform_control/segment.py [hunk 12]
```diff
@@ -810,15 +836,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
...
+        Routine to properly combine elements in the segment.
+        :param elements: list or set of elements in the segment to be combined
+        :param combined_el_name: name of the combined element
+        :return:
+        """
+
+        new_pulse_list = []
+
+        for el in elements:
+            new_pulse_list+= self.elements.pop(el)
```
Comment (antsr) [inline pycqed/measurement/waveform_control/segment.py]: space missing

File: pycqed/measurement/waveform_control/segment.py [hunk 13]
```diff
@@ -164,6 +182,8 @@ def resolve_segment(self, allow_overlap=False,
         self.resolve_mirror()
         self.resolve_Z_gates()
         self.add_flux_crosstalk_cancellation_channels()
+        if self.resolve_overlapping_elements:
+            self.resolve_overlap()
         self.gen_trigger_el(allow_overlap=allow_overlap)
```
Comment (antsr) [inline pycqed/measurement/waveform_control/segment.py:203]: Since the trigger elements are generated after resolving overlap, we might create new overlaps in the process.
Since we currently generate all trigger pulses from the master AWG, it is not an immediate issue, but could catch us in the future if we want to trigger from non-master AWGs.
I think a visible comment should be added to the documentation of the `resolve_overlapping_elements` parameters to remind us of this.

**Review by antsr — APPROVED** [review]
Ok to merge by me when comments from me and Christoph have been adressed.

File: pycqed/measurement/waveform_control/segment.py (hunk 10 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Done

File: pycqed/measurement/waveform_control/segment.py (hunk 1 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py:108]: It is documented in line 47, or what do you mean?

File: pycqed/measurement/waveform_control/segment.py (hunk 13 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py:203]: Added a NB in the documentation of the pulsar parameter.

File: pycqed/measurement/waveform_control/segment.py (hunk 2 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py:862]: Indeed this is a very good comment. When writing this I wanted to reuse the code that we already use for testing whether two elements overlap. If the working assumption is, that elements do not overlap, this criterion is sufficient for detecting *one* overlapping element. As you noticed, it does not work if the next element overlaps with the second to last one (or elements before that).  

I do see to solutions to this problem:
1.) when finding an overlapping element, test whether this element also overlaps with any of the following elements. 
2.) use an entirely different approach for testing for overlaps, trying to test overlaps of all elements with any of the other elements. 

While approach 2.) is more elegant in the sense that it is more general (less if statements), it might be computationally more intense (scales with num_el**2). Especially since the general case will still be to not have overlapping elements, such a test is not required in most instances. 

I will go ahead with testing 1.) (as it is also easier to implement). Curious to hear your opinion @chellings

Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py:862]: I pushed my suggested solution (I tested it for your suggested case and also the other cases I tested before). Let me know what you think @chellings .

File: pycqed/measurement/waveform_control/segment.py (hunk 3 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Done.

File: pycqed/measurement/waveform_control/segment.py (hunk 11 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Done.

File: pycqed/measurement/waveform_control/segment.py (hunk 12 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Done.

File: pycqed/measurement/waveform_control/segment.py (hunk 7 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Done.

File: pycqed/measurement/waveform_control/segment.py (hunk 8 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py:958]: Done.

File: pycqed/measurement/waveform_control/segment.py (hunk 6 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Np.inf can be used (I used notation that was used further up in the code, but changed it at both positions. 

Added comment.

File: pycqed/measurement/waveform_control/segment.py (hunk 4 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: I think you are right, that it is sufficient to only test whether it it in the last entry of `joint_overlapping_elements`. I will test it like this, even though it is always difficult to come up with edge cases for which this might not work.

Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Seems to work, nice catch!

File: pycqed/measurement/waveform_control/segment.py (hunk 5 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: I like the above suggestion. Implemented it like you suggested.

File: pycqed/measurement/waveform_control/segment.py (hunk 9 above)
Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: I cannot see it unfortunately. Maybe we can briefly talk.

**Comment by MichKe95** [comment]
Done implementing (most of) your comments. Once I could address the remaining questions I have with @chellings it should be good to merge. 

Thanks for your thorough review!

File: pycqed/measurement/waveform_control/segment.py (hunk 1 above)
Comment (chellings) [reply pycqed/measurement/waveform_control/segment.py:108]: haha, you added line 47 with a commit that you pushed after my review

File: pycqed/measurement/waveform_control/segment.py (hunk 2 above)
Comment (chellings) [reply pycqed/measurement/waveform_control/segment.py:862]: Fine by me, the suggested solution sounds reasonable. I can have a final quick look at the code ASAP.

File: pycqed/measurement/waveform_control/segment.py (hunk 9 above)
Comment (chellings) [reply pycqed/measurement/waveform_control/segment.py]: Each entry in your new_pulse_list is an UnresolvedPulse object, which has a pulse_obj property, see the init of UnresolvedPulse. This pulse_obj is a Pulse object, and thus has a property element_name, see the init of Pulse in pycqed.measurement.waveform_control.pulse. This property element_name will be outdated after your code.

Comment (MichKe95) [reply pycqed/measurement/waveform_control/segment.py]: Ah yes, now I see it and makes sense! (self.elements is however a list of pulse objects already after resolving, so updating the p.element_name directly is sufficient).

**Comment by MichKe95** [comment]
I believe this was the last comment I had to address! Good to merge @chellings ?

**Review by chellings — APPROVED** [review]
Thanks for the final improvements!
